### PR TITLE
Properly close files

### PR DESCRIPTION
**Pull request by wryun**
Thanks for picking me up - I'm usually good about this in Python, but apparently Go brings out my worst habits :)

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -103,25 +103,37 @@ func run(l *log.Logger, args arguments) error {
...
+
+		defer f.Close()
+
+		input = f
 	}
 
-	var input io.Reader
-	if args.templateFile == "-" {
-		input = os.Stdin
+	var out io.Writer
```
Comment (wryun) [inline main.go]: switched the order so we fail without writing anything if there's a borked input file.

File: main.go [hunk 2]
```diff
@@ -103,25 +103,37 @@ func run(l *log.Logger, args arguments) error {
...
+		input = os.Stdin
 	} else {
-		var err error
-		out, err = os.Create(args.outputFile)
+		f, err := os.Open(args.templateFile)
 		if err != nil {
 			return err
 		}
+
+		defer f.Close()
```
Comment (wryun) [inline main.go]: I don't report the error here - standard library doesn't either (in ioutil.ReadFile). It's pretty meaningless, I think, if we've managed to _read_ as much as we want successfully?

File: main.go [hunk 3]
```diff
@@ -103,25 +103,37 @@ func run(l *log.Logger, args arguments) error {
...
 	} else {
-		input, err = os.Open(args.templateFile)
+		f, err := os.Create(args.outputFile)
 		if err != nil {
 			return err
 		}
+
+		defer func() {
+			err := f.Close()
+			if err != nil {
```
Comment (ash2k) [inline main.go]: It is better to keep the first error if it is there already. Change to `if finalError == nil {`?

Comment (ash2k) [reply main.go]: `defer encoder.Close()` also returns an error and should be handled the same way. You could write a tiny helper func that takes a pointer to `finalError` and an `io.Closer` to make it all slightly less messy.

File: main.go (hunk 2 above)
Comment (ash2k) [reply main.go]: Yes, I think this is fine.

**Review by ash2k — APPROVED** [review]

**Pull request merged**